presto-router/src/main/java/com/facebook/presto/router/cluster/ClusterManager.java

```diff
@@ -21,88 +22,372 @@
+                    remoteClusterInfos.remove(uri);
+                    remoteQueryInfos.remove(uri);
+                    log.info("Successfully removed cluster %s from the router", uri.getHost());
+                }
+            }
+            this.currentConfig.set(new ClusterManagerConfig(newGroups, newGroupSelectors, newScheduler, newSchedulerType));
+        };
+        onConfigChangeDetection.apply();
+    }
 
-        this.groups = ImmutableMap.copyOf(routerSpec.getGroups().stream().collect(toMap(GroupSpec::getName, group -> group)));
-        this.groupSelectors = ImmutableList.copyOf(routerSpec.getSelectors());
-        this.schedulerType = routerSpec.getSchedulerType();
-        this.scheduler = new SchedulerFactory(routerSpec.getSchedulerType()).create();
+    @PostConstruct
+    public void startConfigReloadTaskFileWatcher()
+    {
+        CompletableFuture.supplyAsync(() -> {
+            try (WatchService watchService = FileSystems.getDefault().newWatchService()) {
+                File routerConfigFile = new File(routerConfig.getConfigFile());
+                log.info(String.format("Router config watch service monitoring %s", routerConfig.getConfigFile()));
+                Path parentDir = routerConfigFile.toPath().getParent();
+                parentDir.register(
+                        watchService,
+                        new WatchEvent.Kind[] {
+                                StandardWatchEventKinds.ENTRY_MODIFY,
+                                StandardWatchEventKinds.ENTRY_CREATE,
+                                StandardWatchEventKinds.ENTRY_DELETE,
+                                StandardWatchEventKinds.OVERFLOW},
+                        SensitivityWatchEventModifier.HIGH);
+                isWatchServiceStarted.set(true);
+                log.info("Successfully registered watch service for %s", parentDir);
 
-        this.initializeServerWeights();
+                while (true) {
+                    WatchKey key = watchService.take();
+                    log.info("Changes to router config directory detected: %s", routerConfigFile);
+                    for (WatchEvent<?> event : key.pollEvents()) {
+                        log.info("Event detected: %s, path: %s", event.kind().name(), event.context());
+                        Path changed = (Path) event.context();
+                        if (changed.endsWith(routerConfigFile.getName())) {
+                            this.onConfigChangeDetection.apply();
+                        }
+                        else {
+                            log.info("Config change to %s ignored by ClusterManager (config file is %s)", event.context(), routerConfigFile.getName());
+                        }
+                    }
+                    key.reset();
+                }
+            }
+            catch (Exception e) {
+                log.error("Exception in file watcher loop while monitoring %s, %s", routerConfig.getConfigFile(), e);
```

Comment:
If this future exits for some reason, should we stop the router? or attempt to restart it?

Reply:
I think you can inject a `LifeCycleManager` and call `lifecycleManager.stop()` in the case this loop ever exits

Reply:
The other option, is to make sure that the loop has no exit conditions by catching all exceptions and only breaking on an interrupt